pkg/querier/remote_read.go

```diff
@@ -6,78 +6,159 @@
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
 		}
-		errors := make(chan error)
-		for i, qr := range req.Queries {
-			go func(i int, qr *client.QueryRequest) {
-				from, to, matchers, err := client.FromQueryRequest(qr)
-				if err != nil {
-					errors <- err
-					return
-				}
-
-				querier, err := q.Querier(ctx, int64(from), int64(to))
-				if err != nil {
-					errors <- err
-					return
-				}
-
-				params := &storage.SelectHints{
-					Start: int64(from),
-					End:   int64(to),
-				}
-				seriesSet := querier.Select(false, params, matchers...)
-				resp.Results[i], err = seriesSetToQueryResponse(seriesSet)
-				errors <- err
-			}(i, qr)
+
+		switch respType {
+		case client.STREAMED_XOR_CHUNKS:
+			remoteReadStreamedXORChunks(ctx, q, w, &req, maxBytesInFrame, logger)
+		default:
+			remoteReadSamples(ctx, q, w, &req, logger)
 		}
+	})
+}
+
+func remoteReadSamples(
+	ctx context.Context,
+	q storage.Queryable,
+	w http.ResponseWriter,
+	req *client.ReadRequest,
+	logger log.Logger,
+) {
+	resp := client.ReadResponse{
+		Results: make([]*client.QueryResponse, len(req.Queries)),
+	}
+	// Fetch samples for all queries in parallel.
+	errCh := make(chan error)
 
-		var lastErr error
-		for range req.Queries {
-			err := <-errors
+	for i, qr := range req.Queries {
+		go func(i int, qr *client.QueryRequest) {
+			from, to, matchers, err := client.FromQueryRequest(qr)
 			if err != nil {
-				lastErr = err
+				errCh <- err
+				return
 			}
+
+			querier, err := q.Querier(ctx, int64(from), int64(to))
+			if err != nil {
+				errCh <- err
+				return
+			}
+
+			params := &storage.SelectHints{
+				Start: int64(from),
+				End:   int64(to),
+			}
+			seriesSet := querier.Select(false, params, matchers...)
+			resp.Results[i], err = seriesSetToQueryResponse(seriesSet)
+			errCh <- err
+		}(i, qr)
+	}
+
+	var lastErr error
+	for range req.Queries {
+		err := <-errCh
+		if err != nil {
+			lastErr = err
 		}
-		if lastErr != nil {
-			http.Error(w, lastErr.Error(), http.StatusBadRequest)
+	}
+	if lastErr != nil {
+		http.Error(w, lastErr.Error(), http.StatusBadRequest)
+		return
+	}
+	w.Header().Add("Content-Type", "application/x-protobuf")
+	w.Header().Set("Content-Encoding", "snappy")
+
+	if err := util.SerializeProtoResponse(w, &resp, util.RawSnappy); err != nil {
+		level.Error(logger).Log("msg", "error sending remote read response", "err", err)
+	}
+}
+
+func remoteReadStreamedXORChunks(
+	ctx context.Context,
+	q storage.ChunkQueryable,
+	w http.ResponseWriter,
+	req *client.ReadRequest,
+	maxBytesInFrame int,
+	logger log.Logger,
+) {
+	f, ok := w.(http.Flusher)
+	if !ok {
+		http.Error(w, "internal http.ResponseWriter does not implement http.Flusher interface", http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse")
+
+	for i, qr := range req.Queries {
+		if err := func() error {
```

Comment:
That's a funny implementation... we define function inside if statement, then call it, and get error that we then compare in the if condition. Can we make it little less obscure?

Reply:
Fair enough. 👍 

Changed @ 11dfca178000849f7a18bb8c3d18c40cfef98b1d